From d54d978a7f0d525ba6ea5871e317e5865120e796 Mon Sep 17 00:00:00 2001
From: Tom Stellard <tstellar@redhat.com>
Date: Sat, 29 Apr 2017 02:03:23 +0000
Subject: [llvm] CMake: Split static library exports into their own export file

Summary:
This is to better support distros which split the static libraries into
their own package.

The current problem is that any project the includes LLVMConfig.cmake
will fail to configure unless the static libraries are installed.  This
is because LLVMConfig.cmake includes LLVMExports.cmake, which throws an
error if it can't find files linked to one of the exported targets.

This patch resolves the problem by putting the static library targets
into their own export file, LLVMStaticExports.cmake.  This file
is optionally included by LLVMConfig.cmake, so distros can put this
new file in their static library package to make  LLVMConfig.cmake
no longer depend on these libraries when they are not installed.

Reviewers: beanz, mgorny, chapuni

Subscribers: llvm-commits

Differential Revision: https://reviews.llvm.org/D32668
---
 llvm/cmake/modules/AddLLVM.cmake       | 6 +++++-
 llvm/cmake/modules/CMakeLists.txt      | 3 +++
 llvm/cmake/modules/LLVMConfig.cmake.in | 2 ++
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git llvm/cmake/modules/AddLLVM.cmake llvm/cmake/modules/AddLLVM.cmake
index b74adc11ade9..74a07117918a 100644
--- llvm/cmake/modules/AddLLVM.cmake
+++ llvm/cmake/modules/AddLLVM.cmake
@@ -760,7 +760,11 @@ macro(add_llvm_library name)
       if(${name} IN_LIST LLVM_DISTRIBUTION_COMPONENTS OR
           (in_llvm_libs AND "llvm-libraries" IN_LIST LLVM_DISTRIBUTION_COMPONENTS) OR
           NOT LLVM_DISTRIBUTION_COMPONENTS)
-        set(export_to_llvmexports EXPORT LLVMExports)
+        if (ARG_SHARED)
+          set(export_to_llvmexports EXPORT LLVMExports)
+        else()
+          set(export_to_llvmexports EXPORT LLVMStaticExports)
+        endif()
         set_property(GLOBAL PROPERTY LLVM_HAS_EXPORTS True)
       endif()
 
diff --git llvm/cmake/modules/CMakeLists.txt llvm/cmake/modules/CMakeLists.txt
index 4b8879f65fe4..0b367ab8d859 100644
--- llvm/cmake/modules/CMakeLists.txt
+++ llvm/cmake/modules/CMakeLists.txt
@@ -123,6 +123,7 @@ endif()
 
 set(LLVM_CONFIG_EXPORTS_FILE "\${LLVM_CMAKE_DIR}/LLVMExports.cmake")
 set(LLVM_CONFIG_EXPORTS "${LLVM_EXPORTS}")
+set(LLVM_CONFIG_STATIC_EXPORTS_FILE "\${LLVM_CMAKE_DIR}/LLVMStaticExports.cmake")
 configure_file(
   LLVMConfig.cmake.in
   ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/LLVMConfig.cmake
@@ -139,6 +140,8 @@ if (NOT LLVM_INSTALL_TOOLCHAIN_ONLY)
   if(llvm_has_exports)
     install(EXPORT LLVMExports DESTINATION ${LLVM_INSTALL_PACKAGE_DIR}
             COMPONENT cmake-exports)
+    install(EXPORT LLVMStaticExports DESTINATION ${LLVM_INSTALL_PACKAGE_DIR}
+            COMPONENT cmake-exports)
   endif()
 
   install(FILES
diff --git llvm/cmake/modules/LLVMConfig.cmake.in llvm/cmake/modules/LLVMConfig.cmake.in
index 4d8e33711d27..ffd224f9cecf 100644
--- llvm/cmake/modules/LLVMConfig.cmake.in
+++ llvm/cmake/modules/LLVMConfig.cmake.in
@@ -103,6 +103,8 @@ if(NOT TARGET LLVMSupport)
   set(LLVM_EXPORTED_TARGETS "@LLVM_CONFIG_EXPORTS@")
   include("@LLVM_CONFIG_EXPORTS_FILE@")
   @llvm_config_include_buildtree_only_exports@
+
+  include("@LLVM_CONFIG_STATIC_EXPORTS_FILE@" OPTIONAL)
 endif()
 
 # By creating intrinsics_gen, omp_gen and acc_gen here, subprojects that depend
-- 
2.31.1

